INSTALL.md

```diff
@@ -1,76 +1,217 @@
-# Using `streamly` package
+# Getting started with the `streamly` package
 
-If you are new to Haskell please take a look at [Haskell Getting Started
-Guide](https://github.com/composewell/haskell-dev/blob/master/getting-started.rst).
+This guide will show you how to write a simple program that uses
+`streamly`, using the [`cabal`](https://www.haskell.org/cabal/) build
+manager for Haskell.
 
-The following instructions assume that you are using a shell on a POSIX
-system. On Windows you need a shell under `msys`, some of the commands
-may have slight differences.
+<!-- TODO: Add instructions for `stack` and `nix`.
 
-## Which version to use?
+If you are using `stack` or `nix` please make sure to add the latest
+version from Hackage to your tool configuration.  -->
 
-* [Latest stable release](https://hackage.haskell.org/package/streamly) is available on
-  Hackage.
-* [Latest development version](https://github.com/composewell/streamly) is
-  available on github.
+No prior knowledge of either Haskell or of `cabal` is needed.  We do
+however assume that you are using a command-line shell on a POSIX
+operating system.  If you are running Windows&trade; then you may need
+to run a command-line shell under `msys` and some of the commands
+below may also need to be changed in small ways to make them work.
 
-If you are using `stack` or `nix` please make sure to add the latest version
-from Hackage to your tool configuration.
+**Note**: If you are new to Haskell you may find the [Haskell Getting
+Started
+Guide](https://github.com/composewell/haskell-dev/blob/master/getting-started.rst)
+to be useful.
 
-## Recommended build options
+## Getting started with `streamly` using `cabal`
 
-See [recommended compilation options here](docs/Build.md).
+To get started, you will need a recent `cabal` binary to be installed
+on your system.
 
-## Using with `cabal`
+### Installing (and updating) `cabal`
```

Comment:
I am wondering if this section has gotten too big and a bit distracting for this guide? An alternative is to move this to the "Haskell getting started guide" and put a reference to that section here? I know that guide may also need some rework to fit it there.

Reply:
Yes there is some degree of overlap currently.

For this rewrite, I assumed that the reader of this guide is likely to be curious about `streamly` but also new to Haskell -- since, statistically speaking, that is likely to be the case today. The idea then was offer this new user a simple, 'linear', series of steps that they could follow to get `streamly` up and running quickly on their system.

The 'Getting Started' guide is a great resource, but it covers more topics than strictly needed to bring up a `streamly` environment.  This means that the burden of sifting out the parts of the document that are relevant to bringing up `streamly` would then fall onto the reader, who may or may not know enough to decide.

One way out could be to split the 'Getting Started' guide into focussed sections that could perhaps be referenced from this guide.

Even better perhaps could be to offer a `stack` template that offers everything preconfigured.  With this approach, new users could get started using `streamly` with just two steps:

1. Install `stack`.
2. Run `stack new project-name streamly-template`.

Reply:
> One way out could be to split the 'Getting Started' guide into focussed sections that could perhaps be referenced from this guide.

Yes, that's what I meant. But we can keep it the way you have written for now to not get distracted. I agree with all the previous paragraphs as well.

I think one complication here is because of the cabal upgrade instructions. If we install cabal directly then we do not need upgrade. To make the install simple I have used the `ghcup` method as the preferred method in the Haskell guide. It installs latest versions of cabal and ghc both in one simple step.

I like the direct method instead of using templates because otherwise the users have to be aware of what the templates are, how they work, what exactly they will do etc. And templates are not very useful beyond beginners. If we keep that out of the way, then in both cabal and stack the process is pretty similar except that in cabal you have to install ghc separately:

* install cabal/ghc, stack
* cabal/stack init/new
* ...the rest

If we use `ghcup` to install cabal/ghc then both are equivalent.

Reply:
> If we install cabal directly then we do not need upgrade ...

True.  However, the downside of these "direct" installation methods is that they require users to download and run pre-compiled binaries of unknown provenance.

Reply:
I guess stack install is similar in nature to ghcup and cabal binary download. So all of them should suffer from the same security issues. I am wondering if you trust the website, are using https and verifying the signatures then is it sufficient?

Reply:
I believe `stack` offers a mode for upgrading from source code (`stack upgrade --source-only`).

I noticed that the installation instructions for a number of new installers (`ghcup`, `rustup`, etc) invite the user to directly execute a binary downloaded from an external website.

This means that the user would need to trust that:

1. That the source code used to build that tool was trustworthy.
2. That the source code for the tool reached the compilation toolchain on the development machine unmodified.
3. That the compilation toolchain on that machine had not been tampered with.
4. That the executable that was made available on the website was exactly that which was built.
5. That the executable that reached the local disk was exactly that which was made available on the website.

For open-source code, we can protect against (1) with a manual audit.  For (4) and (5) we can sign the built binary and insist that the user verify its signature before proceeding further. However (2) and (3) to defend against.

Building from source reduces the attack surface somewhat, although we would still need to trust that the versions of `stack`, `cabal`, `ghc` and the rest of the toolchain in the default package repository had been bootstrapped correctly.

Reply:
Trust is hard and at odds with convenience. For example, (1) involves a very high cost, you need someone who is competent and trustworthy to do it. The rest can be automated using a trust chain (perhaps only in theory as of now) using the right tools which are themselves verified for trust. But then again there could be tricky issues like the possibility of code injection if the code is running on an untrusted machine. I guess we can stop at a point where the ROI is sufficiently low.